docs/CppInteroperability/ForwardVision.md

```diff
@@ -0,0 +1,266 @@
+# Using C++ from Swift 
+
+## Introduction
+
+This document lays out a vision for the development of the "forward" half of C++ and Swift interoperability: using C++ APIs from Swift. It sets overarching goals that drive the project’s design decisions, outlines some high-level topics related to C++ interoperability, and, finally, investigates a collection of specific API patterns and proposes potential ways for the compiler to import them. This vision is a sketch, rather than a final design for C++ and Swift interoperability. Towards the end, this document suggests a proeccess for evolving C++ interoperability over time, and the path for finalizing the designs discussed here.
+
+“Reverse” interoperability (using Swift APIs from C++) is another extremely important part of the interoperability story; however, reverse interoperability has largely different goals and constraints, which necessarily mean a different design and therefore a different vision document. [The vision for reverse interoperability is being developed here.](https://github.com/apple/swift/pull/61256)
+
+This document is a prospective feature vision document, as described in the [draft review management guidelines](https://github.com/rjmccall/swift-evolution/blob/057b2383102f34c3d0f5b257f82bba0f5b94683d/review_management.md#future-directions-and-roadmaps) of the Swift evolution process.  It has not yet been approved by the Language Workgroup.
+
+## Goals
+
+Safety is a top priority for the Swift programming language, which creates a tension with C++. While Swift enforces strong rules around things like memory safety, mutability, and nullability, C++ largely makes the programmer responsible for handling them correctly, on the pain of undefined behavior. Simply using C++ APIs should not completely undermine Swift's lanaguage guarantees, especially guarantees around safety. At a minimum, imported C++ APIs should generally not be less safe to use from Swift than they would be in C++, and C++ interoperability should strive to make imported APIs *safer* in Swift than they are in C++ by providing safe API interfaces for common, unsafe C++ API patterns (such as iterators). When it is possible for the Swift compiler to statically derive safety properties and API semantics (i.e., how to safely use an API in Swift) from the C++ API interface, C++ interoperability should take advantage of this information. When that is not possible, C++ interoperability should provide annotations to communicate the necessary information to use these APIs safely in Swift. When APIs cannot be used safely or need careful management, Swift should make that clear to the programmer. As a last resort, Swift should make an API unavailable if there's no reasonable path to a sufficiently safe Swift interface for it.
```

Comment:
Unfortunately, I think the nature of C++ is that there will be unsafe corner cases. There is always a trade-off between usability and safety, and we are already leaning heavily towards the safety side of things. This originally read "must," but we decided to weaken it, because otherwise we'd simply not import too many APIs (or break compatibility with Objective-C, for which this guarantee is not there). I think it's OK for this to be a bit subjective.

Reply:
> I think the nature of C++ is that there will be unsafe corner cases

I don't think any differently. 🙂 But IMO it should not be possible in _safe_ Swift to introduce unsafety that isn't due to the nature of the underlying C++ API itself. Would it be too restrictive to commit to something like "APIs which cannot be imported without introducing _additional_ safety concerns on the Swift side should be imported with an explicit indicator of that unsafety (such as `Unsafe` in the name)"?

Reply:
I think the two big issues with adding unsafe to potentially unsafe APIs are 1) 90% of the time the APIs are safe, so "unsafe" starts to lose it's meaning. I think we should try to reserve this for the cases where we are more confident that it is actually an unsafe API. 2) this potentially breaks compatibility with Objective-C, or at least makes it more difficult to migrate bridging headers. Looking at the string example from one of the Swift compiler bridging headers:

```
BridgedIdentifier
SwiftASTContext_getIdentifier(
    void *ctx, const uint8_t * _Nullable str, long len);
```

This would definitely be imported as an "Unsafe" API (without more information). Maybe we want to enable C++ interop though, so we can start replacing these "Bridged" types with C++ types. But let's say we want to start using C++ interop incrementally, we can't update the whole bridging header in one go. This is where the issue comes in: if we start re-naming functions like this, it will prevent people from adopting C++ interop incrementally (they'll have to go find-replace all of these function names as soon as they enable C++ interop). Does that make sense?

Reply:
Hm, I don't think I'm suggesting that this function would necessarily have to change to be called `unsafe_SwiftASTContext_getIdentifier` or something—this API would _already_ be imported to Swift with 'unsafe' sigils on its `ctx` and `str` parameters.

Further, I'm not saying that _every_ potentially unsafe C++ API should be imported to Swift with 'unsafe' markings, only those for which importing to Swift introduces _new_ forms of unsafety that would not be possible to achieve in C++. I.e., the APIs that we cannot guarantee are "[at least as] safe to use from Swift [as] they would be in C++." Is this `SwiftASTContext_getIdentifier` an API that would fall into that bucket, you think?